gprofiler/profiler_base.py

```diff
@@ -75,3 +79,46 @@ class NoopProfiler(ProfilerInterface):
 
     def snapshot(self) -> ProcessToStackSampleCounters:
         return {}
+
+
+class ProcessProfilerBase(ProfilerBase):
+    """
+    Base class for process-based profilers: those that operate on each process separately, thus need
+    to be invoked for each PID.
+    This class implements snapshot() for them - creates a thread that runs _profile_process() for each
+    process that we wish to profile; then waits for all and returns the result.
+    """
+
+    def _select_processes_to_profile(self) -> List[Process]:
+        raise NotImplementedError
+
+    def _profile_process(self, process: Process) -> Optional[StackToSampleCount]:
+        raise NotImplementedError
+
+    def snapshot(self) -> ProcessToStackSampleCounters:
+        processes_to_profile = self._select_processes_to_profile()
+        if not processes_to_profile:
+            return {}
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=len(processes_to_profile)) as executor:
+            futures = {}
+            for process in processes_to_profile:
+                futures[executor.submit(self._profile_process, process)] = process.pid
+
+            results = {}
+            for future in concurrent.futures.as_completed(futures):
+                try:
+                    result = future.result()
+                    if result is not None:
+                        results[futures[future]] = result
+                except StopEventSetException:
+                    raise
```

Comment:
Consider raising this after calling `.result()` on all of the threads to make sure you are not losing exceptions

Reply:
That's true - we are aware of the issue, will be solved separately (there are other cases as well)